### PR TITLE
Enable planner to be used for loading sharded optimizer state dict

### PR DESCRIPTION
**Pull request by b-chu**
This creates a more consistent interface for saving and loading sharded state dicts. A planner is able to be specified when saving a sharded optimizer state dict, but there is currently no planner support for loading one. This change does not affect the default behavior of the function.

### CODE REVIEW

**Comment by fegin** [comment]
`load_sharded_optimizer_state_dict` is being deprecated in favor of the new state_dict APIs, `torch.distributed.checkpoint.state_dict.get_state_dict` and `torch.distributed_checkpoint.state_dict.set_state_dict`. We can still land this PR to unblock the existing use cases. cc., @wz337

**Comment by wz337** [comment]
@b-chu As Chien-Chin mentioned, we are going to deprecate this API in favor of `torch.distributed.checkpoint.state_dict.get_state_dict`, as this would initialize the optimizer state ahead of time so optimizer loading would not be handled separately. 

For more reference, you can check our slides here: https://static.sched.com/hosted_files/pytorch2023/00/%5BPTC%2023%5D%20Distributed%20Checkpoint.pdf

**Review by wz337 — APPROVED** [review]
LGTM. Accept to unblock existing use case.

**Comment by b-chu** [comment]
@fegin @wz337 would you mind adding the appropriate labels?

**Comment by ZainRizvi** [comment]
@pytrochbot revert -m "Sorry, but this breaks internal builds. @wz337 can you please help fix this?" -c ghfirst

**Comment by wz337** [comment]
> @pytorchbot revert -m "Sorry, but this breaks internal builds. @wz337 can you please help fix this?" -c ghfirst

@ZainRizvi  How could I get more information regarding the internal building that is broken, since I don't see any unit test failure in CI?
